src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,894 @@
+  const dfcc_loop_nesting_grapht &loop_nesting_graph)
+{
+  std::list<size_t> must_have_contracts;
+  for(std::size_t idx = 0; idx < loop_nesting_graph.size(); idx++)
+  {
+    // keep track of which loops have contracts
+    const auto &node = loop_nesting_graph[idx];
+    if(has_contract(node.latch))
+    {
+      for(const auto pred_idx : loop_nesting_graph.get_predecessors(idx))
+        must_have_contracts.push_back(pred_idx);
+    }
+  }
+
+  while(!must_have_contracts.empty())
+  {
+    const auto idx = must_have_contracts.front();
+    must_have_contracts.pop_front();
+    const auto &node = loop_nesting_graph[idx];
+    if(!has_contract(node.latch))
+    {
+      throw invalid_source_file_exceptiont(
+        "Found loop without contract nested in a loop with a contract.\nPlease "
+        "provide a contract or unwind this loop before applying loop "
+        "contracts.",
+        node.head->source_location());
+    }
+
+    for(const auto pred_idx : loop_nesting_graph.get_predecessors(idx))
+      must_have_contracts.push_back(pred_idx);
+  }
+}
+
+/// Tags instructions of loops found in \p loop_nesting_graph with the loop
+/// identifier of the innermost loop it belongs to, and the loop instruction
+/// type : head, body, exiting or latch.
+static void tag_loop_instuctions(
+  goto_programt &goto_program,
+  dfcc_loop_nesting_grapht &loop_nesting_graph)
+{
+  for(const auto &idx : loop_nesting_graph.topsort())
+  {
+    auto &node = loop_nesting_graph[idx];
+    auto &head = node.head;
+    auto &latch = node.latch;
+    auto &instructions = node.instructions;
+
+    dfcc_set_loop_head(head);
+    dfcc_set_loop_latch(latch);
+
+    for(auto t : instructions)
+    {
+      // Skip instructions that are already tagged and belong to inner loops.
+      auto loop_id_opt = dfcc_get_loop_id(t);
+      if(loop_id_opt.has_value() && loop_id_opt.value() != idx)
+        continue;
+
+      dfcc_set_loop_id(t, idx);
+
+      if(t != head && t != latch)
+        dfcc_set_loop_body(t);
+
+      if(t->is_goto() && !instructions.contains(t->get_target()))
+      {
+        dfcc_set_loop_exiting(t);
+      }
+    }
+  }
+
+  auto top_level_id = loop_nesting_graph.size();
+
+  // tag remaining instructions as top level
+  for(auto target = goto_program.instructions.begin();
+      target != goto_program.instructions.end();
+      target++)
+  {
+    // Skip instructions that are already tagged (belong to inner loops).
+    auto loop_id_opt = dfcc_get_loop_id(target);
+    if(loop_id_opt.has_value() && loop_id_opt.value() != top_level_id)
+    {
+      continue;
+    }
+    dfcc_set_loop_id(target, top_level_id);
+    dfcc_set_loop_top_level(target);
+  }
+}
+
+/// Utility class that generates a \ref dfcc_loop_infot struct for each loop
+/// of the given \p loop_nesting_graph and inserts them into \p loop_info_map.
+/// This class is used by the constructor of \ref dfcc_cfg_infot.
+class gen_dfcc_loop_infot
+{
```

Comment:
Wouldn't be cleaner to move this to the `.h` file as well (even tho it's only an utility class)?

Reply:
I'd like to keep this private as it is not meant to be used outside of the module